### PR TITLE
[Dataset/CAPI] implement dataset ctor

### PR DESCRIPTION
**Pull request by zhoonit**
- [Dataset/CAPI] implement dataset ctor

```
**Changes proposed in this PR:**
- Implement `ml_train_dataset_create()`
- Implement `ml_train_add_generator()`
- Implement `ml_train_add_file()`
- Add test as per the implementation.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

File: api/capi/src/nntrainer.cpp [hunk 1]
```diff
@@ -138,14 +138,19 @@ static int ml_train_dataset_create(ml_train_dataset_h *dataset,
...
+  }
 
   ml_train_dataset *nndataset = new ml_train_dataset;
   nndataset->magic = ML_NNTRAINER_MAGIC;
   nndataset->in_use = false;
 
   returnable f = [&]() {
-    nndataset->dataset[ML_TRAIN_DATASET_DATA_USAGE_TRAIN] =
-      ml::train::createDataset(type, train);
+    if (train != nullptr) {
```
Comment (kparichay) [inline api/capi/src/nntrainer.cpp:150]: train is a must as far as I know.

Comment (zhoonit) [reply api/capi/src/nntrainer.cpp:150]: Yes, before delegating this function it is checked. This is to share the logic between making an empty dataset and the old way.

Comment (kparichay) [reply api/capi/src/nntrainer.cpp:150]: aah I missed that.

**Review by kparichay — APPROVED** [review]
LGTM

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**